File: PlayerStatsRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using UnityEngine;

namespace BattleFleet.Database
{
    /// <summary>
    /// PlayerStats Repository - CRUD and utility operations for PlayerStats table
    /// Handles all player statistics and progression management
    /// </summary>
    public class PlayerStatsRepository
    {
        private DatabaseManager dbManager;

        public PlayerStatsRepository()
        {
            dbManager = DatabaseManager.Instance;
        }

        /// <summary>
        /// READ - Get stats for a specific player
        /// </summary>
        public PlayerStats GetPlayerStats(int playerID)
        {
            try
            {
                string query = "SELECT * FROM PlayerStats WHERE PlayerID = @playerID";
                Dictionary<string, object> parameters = new Dictionary<string, object>
                {
                    { "@playerID", playerID }
                };

                IDataReader reader = dbManager.ExecuteReader(query, parameters);
                
                if (reader != null && reader.Read())
                {
                    PlayerStats stats = new PlayerStats
                    {
                        StatID = Convert.ToInt32(reader["StatID"]),
                        PlayerID = Convert.ToInt32(reader["PlayerID"]),
                        Level = Convert.ToInt32(reader["Level"]),
                        TotalExperience = Convert.ToInt32(reader["TotalExperience"]),
                        TotalBattles = Convert.ToInt32(reader["TotalBattles"]),
                        TotalWins = Convert.ToInt32(reader["TotalWins"]),
                        TotalLosses = Convert.ToInt32(reader["TotalLosses"]),
                        TotalDraws = Convert.ToInt32(reader["TotalDraws"]),
                        AverageAccuracy = Convert.ToDouble(reader["AverageAccuracy"]),
                        TotalShipsDestroyed = Convert.ToInt32(reader["TotalShipsDestroyed"]),
                        TotalShipsLost = Convert.ToInt32(reader["TotalShipsLost"]),
                        CampaignProgressPercentage = Convert.ToDouble(reader["CampaignProgressPercentage"]),
                        HighestLevel = Convert.ToInt32(reader["HighestLevel"]),
                        LastUpdated = Convert.ToDateTime(reader["LastUpdated"])
                    };

                    reader.Close();
                    return stats;
                }

                if (reader != null) reader.Close();
                return null;
            }
            catch (Exception e)
            {
                Debug.LogError($"Error retrieving player stats: {e.Message}");
                return null;
            }
        }

        /// <summary>
        /// UPDATE - Add experience to player
        /// Handles level-up logic automatically
        /// </summary>
        public bool AddExperience(int playerID, int experiencePoints)
        {
            try
            {
                PlayerStats stats = GetPlayerStats(playerID);
                if (stats == null)
                {
                    Debug.LogError($"Player stats not found for player {playerID}");
                    return false;
                }

                stats.TotalExperience += experiencePoints;

                // Level up logic: 1000 XP per level
                int previousLevel = stats.Level;
                int experiencePerLevel = 1000;
                stats.Level = (stats.TotalExperience / experiencePerLevel) + 1;

                // Track highest level reached
                if (stats.Level > stats.HighestLevel)
                {
                    stats.HighestLevel = stats.Level;
                }

                // If leveled up, log it
                if (stats.Level > previousLevel)
                {
                    Debug.Log($"Player {playerID} leveled up! Level: {previousLevel} → {stats.Level}");
                }

                return UpdatePlayerStats(stats);
            }
            catch (Exception e)
            {
                Debug.LogError($"Error adding experience: {e.Message}");
                return false;
            }
        }

        /// <summary>
        /// UPDATE - Record a battle result and update stats accordingly
        /// </summary>
        public bool RecordBattleResult(int playerID, BattleRecord battle)
        {
            try
            {
                PlayerStats stats = GetPlayerStats(playerID);
                if (stats == null)
                {
                    Debug.LogError($"Player stats not found for player {playerID}");
                    return false;
                }

                // Update battle records
                stats.TotalBattles++;

                if (battle.Result == "Win")
                {
                    stats.TotalWins++;
                }
                else if (battle.Result == "Loss")
                {
                    stats.TotalLosses++;
                }
                else if (battle.Result == "Draw")
                {
                    stats.TotalDraws++;
                }

                // Update ship statistics
                stats.TotalShipsDestroyed += battle.ShipsDestroyed;
                stats.TotalShipsLost += battle.ShipsLost;

                // Add experience gained from battle
                stats.TotalExperience += battle.ExperienceGained;

                // Recalculate level
                int experiencePerLevel = 1000;
                int previousLevel = stats.Level;
                stats.Level = (stats.TotalExperience / experiencePerLevel) + 1;

                if (stats.Level > stats.HighestLevel)
                {
                    stats.HighestLevel = stats.Level;
                }

                if (stats.Level > previousLevel)
                {
                    Debug.Log($"Player {playerID} leveled up! Level: {previousLevel} → {stats.Level}");
                }

                // Update last modified time
                stats.LastUpdated = DateTime.Now;

                return UpdatePlayerStats(stats);
            }
            catch (Exception e)
            {
                Debug.LogError($"Error recording battle result: {e.Message}");
                return false;
            }
        }

        /// <summary>
        /// UPDATE - Update accuracy statistics
        /// </summary>
        public bool UpdateAccuracy(int playerID, int totalShots, int successfulShots)
        {
            try
            {
                if (totalShots <= 0)
                {
                    Debug.LogWarning("Total shots must be greater than 0");
                    return false;
                }

                PlayerStats stats = GetPlayerStats(playerID);
                if (stats == null)
                {
                    Debug.LogError($"Player stats not found for player {playerID}");
                    return false;
                }

                // Calculate cumulative accuracy
                double currentAccuracyBase = stats.AverageAccuracy * (stats.TotalBattles - 1);
                double newAccuracy = successfulShots / (double)totalShots;
                stats.AverageAccuracy = (currentAccuracyBase + newAccuracy) / stats.TotalBattles;

                return UpdatePlayerStats(stats);
            }
            catch (Exception e)
            {
                Debug.LogError($"Error updating accuracy: {e.Message}");
                return false;
            }
        }

        /// <summary>
        /// UPDATE - Update campaign progress
        /// </summary>
        public bool UpdateCampaignProgress(int playerID, double progressPercentage)
        {
            try
            {
                if (progressPercentage < 0 || progressPercentage > 100)
                {
                    Debug.LogWarning("Progress percentage must be between 0 and 100");
                    return false;
                }

                PlayerStats stats = GetPlayerStats(playerID);
                if (stats == null)
                {
                    Debug.LogError($"Player stats not found for player {playerID}");
                    return false;
                }

                stats.CampaignProgressPercentage = progressPercentage;
                return UpdatePlayerStats(stats);
            }
            catch (Exception e)
            {
                Debug.LogError($"Error updating campaign progress: {e.Message}");
                return false;
            }
        }

        /// <summary>
        /// UPDATE - Update all player stats
        /// </summary>
        private bool UpdatePlayerStats(PlayerStats stats)
        {
            try
            {
                string query = @"UPDATE PlayerStats 
                                SET Level = @level,
                                    TotalExperience = @totalExp,
                                    TotalBattles = @totalBattles,
                                    TotalWins = @totalWins,
                                    TotalLosses = @totalLosses,
                                    TotalDraws = @totalDraws,
                                    AverageAccuracy = @avgAccuracy,
                                    TotalShipsDestroyed = @shipsDestroyed,
                                    TotalShipsLost = @shipsLost,
                                    CampaignProgressPercentage = @campaignProgress,
                                    HighestLevel = @highestLevel,
                                    LastUpdated = @lastUpdated
                                WHERE PlayerID = @playerID";

                Dictionary<string, object> parameters = new Dictionary<string, object>
                {
                    { "@level", stats.Level },
                    { "@totalExp", stats.TotalExperience },
                    { "@totalBattles", stats.TotalBattles },
                    { "@totalWins", stats.TotalWins },
                    { "@totalLosses", stats.TotalLosses },
                    { "@totalDraws", stats.TotalDraws },
                    { "@avgAccuracy", stats.AverageAccuracy },
                    { "@shipsDestroyed", stats.TotalShipsDestroyed },
                    { "@shipsLost", stats.TotalShipsLost },
                    { "@campaignProgress", stats.CampaignProgressPercentage },
                    { "@highestLevel", stats.HighestLevel },
                    { "@lastUpdated", DateTime.Now },
                    { "@playerID", stats.PlayerID }
                };

                int result = dbManager.ExecuteNonQuery(query, parameters);
                return result > 0;
            }
            catch (Exception e)
            {
                Debug.LogError($"Error updating player stats: {e.Message}");
                return false;
            }
        }

        /// <summary>
        /// READ - Get top N players by level
        /// </summary>
        public List<PlayerStats> GetTopPlayersByLevel(int topCount = 10)
        {
            List<PlayerStats> topPlayers = new List<PlayerStats>();

            try
            {
                string query = $"SELECT * FROM PlayerStats ORDER BY Level DESC, TotalExperience DESC LIMIT {topCount}";
                IDataReader reader = dbManager.ExecuteReader(query);

                if (reader != null)
                {
                    while (reader.Read())
                    {
                        PlayerStats stats = new PlayerStats
                        {
                            StatID = Convert.ToInt32(reader["StatID"]),
                            PlayerID = Convert.ToInt32(reader["PlayerID"]),
                            Level = Convert.ToInt32(reader["Level"]),
                            TotalExperience = Convert.ToInt32(reader["TotalExperience"]),
                            TotalBattles = Convert.ToInt32(reader["TotalBattles"]),
                            TotalWins = Convert.ToInt32(reader["TotalWins"]),
                            TotalLosses = Convert.ToInt32(reader["TotalLosses"]),
                            TotalDraws = Convert.ToInt32(reader["TotalDraws"]),
                            AverageAccuracy = Convert.ToDouble(reader["AverageAccuracy"]),
                            TotalShipsDestroyed = Convert.ToInt32(reader["TotalShipsDestroyed"]),
                            TotalShipsLost = Convert.ToInt32(reader["TotalShipsLost"]),
                            HighestLevel = Convert.ToInt32(reader["HighestLevel"])
                        };
                        topPlayers.Add(stats);
                    }
                    reader.Close();
                }
            }
            catch (Exception e)
            {
                Debug.LogError($"Error retrieving top players: {e.Message}");
            }

            return topPlayers;
        }

        /// <summary>
        /// READ - Get top N players by win rate
        /// </summary>
        public List<PlayerStats> GetTopPlayersByWinRate(int topCount = 10)
        {
            List<PlayerStats> topPlayers = new List<PlayerStats>();

            try
            {
                string query = $@"SELECT * FROM PlayerStats 
                                WHERE TotalBattles > 0 
                                ORDER BY (CAST(TotalWins AS FLOAT) / TotalBattles) DESC, TotalBattles DESC 
                                LIMIT {topCount}";
                
                IDataReader reader = dbManager.ExecuteReader(query);

                if (reader != null)
                {
                    while (reader.Read())
                    {
                        PlayerStats stats = new PlayerStats
                        {
                            StatID = Convert.ToInt32(reader["StatID"]),
                            PlayerID = Convert.ToInt32(reader["PlayerID"]),
                            Level = Convert.ToInt32(reader["Level"]),
                            TotalExperience = Convert.ToInt32(reader["TotalExperience"]),
                            TotalBattles = Convert.ToInt32(reader["TotalBattles"]),
                            TotalWins = Convert.ToInt32(reader["TotalWins"]),
                            TotalLosses = Convert.ToInt32(reader["TotalLosses"]),
                            TotalDraws = Convert.ToInt32(reader["TotalDraws"]),
                            AverageAccuracy = Convert.ToDouble(reader["AverageAccuracy"]),
                            TotalShipsDestroyed = Convert.ToInt32(reader["TotalShipsDestroyed"]),
                            TotalShipsLost = Convert.ToInt32(reader["TotalShipsLost"]),
                            HighestLevel = Convert.ToInt32(reader["HighestLevel"])
                        };
                        topPlayers.Add(stats);
                    }
                    reader.Close();
                }
            }
            catch (Exception e)
            {
                Debug.LogError($"Error retrieving top players by win rate: {e.Message}");
            }

            return topPlayers;
        }

        /// <summary>
        /// READ - Get player rank by level
        /// </summary>
        public int GetPlayerRank(int playerID, bool byWinRate = false)
        {
            try
            {
                string query;
                if (byWinRate)
                {
                    query = $@"SELECT COUNT(*) FROM PlayerStats 
                            WHERE (CAST(TotalWins AS FLOAT) / NULLIF(TotalBattles, 0)) > 
                                  (SELECT CAST(TotalWins AS FLOAT) / NULLIF(TotalBattles, 0) FROM PlayerStats WHERE PlayerID = @playerID)";
                }
                else
                {
                    query = $@"SELECT COUNT(*) FROM PlayerStats 
                            WHERE Level > (SELECT Level FROM PlayerStats WHERE PlayerID = @playerID)
                            OR (Level = (SELECT Level FROM PlayerStats WHERE PlayerID = @playerID) 
                                AND TotalExperience > (SELECT TotalExperience FROM PlayerStats WHERE PlayerID = @playerID))";
                }

                Dictionary<string, object> parameters = new Dictionary<string, object>
                {
                    { "@playerID", playerID }
                };

                object result = dbManager.ExecuteScalar(query, parameters);
                return Convert.ToInt32(result) + 1; // +1 because rank is 1-indexed
            }
            catch (Exception e)
            {
                Debug.LogError($"Error getting player rank: {e.Message}");
                return -1;
            }
        }

        /// <summary>
        /// READ - Get experience required for next level
        /// </summary>
        public int GetExperienceForNextLevel(int currentLevel)
        {
            int experiencePerLevel = 1000;
            return (currentLevel) * experiencePerLevel;
        }

        /// <summary>
        /// READ - Calculate remaining experience to next level
        /// </summary>
        public int GetRemainingExperienceForLevel(int playerID)
        {
            try
            {
                PlayerStats stats = GetPlayerStats(playerID);
                if (stats == null) return 0;

                int nextLevelXP = GetExperienceForNextLevel(stats.Level);
                int remaining = nextLevelXP - stats.TotalExperience;
                
                return remaining > 0 ? remaining : 0;
            }
            catch (Exception e)
            {
                Debug.LogError($"Error calculating remaining experience: {e.Message}");
                return 0;
            }
        }

        /// <summary>
        /// RESET - Reset all stats for a player (admin/debug use)
        /// </summary>
        public bool ResetPlayerStats(int playerID)
        {
            try
            {
                string query = @"UPDATE PlayerStats 
                                SET Level = 1,
                                    TotalExperience = 0,
                                    TotalBattles = 0,
                                    TotalWins = 0,
                                    TotalLosses = 0,
                                    TotalDraws = 0,
                                    AverageAccuracy = 0.0,
                                    TotalShipsDestroyed = 0,
                                    TotalShipsLost = 0,
                                    CampaignProgressPercentage = 0.0,
                                    LastUpdated = @lastUpdated
                                WHERE PlayerID = @playerID";

                Dictionary<string, object> parameters = new Dictionary<string, object>
                {
                    { "@playerID", playerID },
                    { "@lastUpdated", DateTime.Now }
                };

                int result = dbManager.ExecuteNonQuery(query, parameters);
                
                if (result > 0)
                {
                    Debug.Log($"Player {playerID} stats reset to default.");
                    return true;
                }

                return false;
            }
            catch (Exception e)
            {
                Debug.LogError($"Error resetting player stats: {e.Message}");
                return false;
            }
        }

        // ── XP & Levelling ─────────────────────────────────────────────────────────
        /// <summary>
        /// Awards XP to a player after a match and handles level-up automatically.
        /// Call this at the end of every game.
        /// Returns the new level (useful to trigger a level-up animation in Unity).
        /// </summary>
        public int AwardXpAndLevelUp(int playerId, int xpGained)
        {
            using var conn = _db.GetConnection();
            conn.Open();

            // 1. Read current state
            using var selectCmd = conn.CreateCommand();
            selectCmd.CommandText = "SELECT fleet_level, fleet_xp FROM players WHERE id = @id;";
            selectCmd.Parameters.AddWithValue("@id", playerId);

            int currentLevel, currentXp;
            using (var reader = selectCmd.ExecuteReader())
            {
                reader.Read();
                currentLevel = reader.GetInt32(0);
                currentXp    = reader.GetInt32(1);
            }

            // 2. Calculate new XP and level
            int newXp    = currentXp + xpGained;
            int newLevel = CalculateLevel(newXp);     // see formula below

            // 3. Persist
            using var updateCmd = conn.CreateCommand();
            updateCmd.CommandText = @"
                UPDATE players
                SET fleet_xp    = @xp,
                    fleet_level = @level
                WHERE id = @id;";
            updateCmd.Parameters.AddWithValue("@xp",    newXp);
            updateCmd.Parameters.AddWithValue("@level",  newLevel);
            updateCmd.Parameters.AddWithValue("@id",     playerId);
            updateCmd.ExecuteNonQuery();

            return newLevel;
        }

        /// <summary>
        /// XP formula: each level requires 100 * level XP on top of the previous.
        /// Level 1 = 0 XP, Level 2 = 100, Level 3 = 300, Level 4 = 600 …
        /// Easy to swap out for any curve you like.
        /// </summary>
        public static int CalculateLevel(int totalXp)
        {
            int level = 1;
            int xpNeeded = 0;
            while (totalXp >= xpNeeded + (level * 100))
            {
                xpNeeded += level * 100;
                level++;
            }
            return level;
        }

        /// <summary>Returns XP needed to reach the next level (for the progress bar).</summary>
        public (int currentXp, int xpForCurrentLevel, int xpForNextLevel) GetProgressBarData(int playerId)
        {
            using var conn = _db.GetConnection();
            conn.Open();

            using var cmd = conn.CreateCommand();
            cmd.CommandText = "SELECT fleet_xp, fleet_level FROM players WHERE id = @id;";
            cmd.Parameters.AddWithValue("@id", playerId);

            using var reader = cmd.ExecuteReader();
            reader.Read();
            int totalXp = reader.GetInt32(0);
            int level   = reader.GetInt32(1);

            // XP thresholds for current level band
            int xpStart = 0;
            for (int l = 1; l < level; l++) xpStart += l * 100;
            int xpEnd = xpStart + level * 100;

            return (totalXp, xpStart, xpEnd);
        }
    }
}

File: database/player_stats_repository.py
# player_stats_repository.py
# Manages PlayerStats (player-level progression) and Fleet-level progression.
#
# Two parallel XP systems exist in this game:
#   1. Player Level  — stored in PlayerStats.Level / TotalExperience
#                      Flat formula: level = total_xp // 1000 + 1
#                      Updated by record_battle_result() and add_experience()
#
#   2. Fleet Level   — stored in Players.FleetLevel / Players.FleetXP
#                      Quadratic formula: each level costs level*100 XP
#                      Updated by award_fleet_xp_and_level_up()
#                      Gates ship unlocks (ShipUnlockRepository reads FleetLevel)
#
# FIX (vs Unity version):
#   - award_fleet_xp_and_level_up() used undefined "_db" field; now uses DatabaseManager
#   - award_fleet_xp_and_level_up() used wrong table "players" / columns "fleet_xp", "fleet_level"
#     (schema uses "Players", "FleetXP", "FleetLevel")
#   - get_progress_bar_data() had the same naming bugs

import logging
from datetime import datetime, timezone
from typing import Callable, Optional

from .database_manager import DatabaseManager
from .models import BattleRecord, PlayerStats

logger = logging.getLogger(__name__)

XP_PER_PLAYER_LEVEL: int = 1000  # Flat curve for the overall player level


class PlayerStatsRepository:
    # GDScript signals become simple callback lists.
    # Register with: repo.on_player_leveled_up.append(my_func)
    # Callback signature: (player_id: int, old_level: int, new_level: int) -> None

    def __init__(self, db: DatabaseManager) -> None:
        self._db = db

    # ── READ ───────────────────────────────────────────────────────────────────

    def get_player_stats(self, player_id: int) -> Optional[PlayerStats]:
        rows = self._db.execute_reader(
            "SELECT * FROM PlayerStats WHERE PlayerID = ?", (player_id,)
        )
        if not rows:
            return None
        return _map_stats(rows[0])


    # ── RECORD BATTLE ──────────────────────────────────────────────────────────

    def record_battle_result(self, player_id: int, battle: BattleRecord) -> bool:
        """Records a battle result, updates wins/losses/experience, recalculates level."""
        stats = self.get_player_stats(player_id)
        if stats is None:
            logger.error("PlayerStatsRepository: Stats not found for player %d.", player_id)
            return False

        stats.total_battles += 1
        if battle.result == "Win":
            stats.total_wins += 1
        elif battle.result == "Loss":
            stats.total_losses += 1
        elif battle.result == "Draw":
            stats.total_draws += 1

        stats.total_ships_destroyed += battle.ships_destroyed
        stats.total_ships_lost += battle.ships_lost

        stats.last_updated = _now()
        self._insert_battle_history(player_id, battle)
        return self._update_player_stats(stats)

    # ── ACCURACY ───────────────────────────────────────────────────────────────

    def update_accuracy(
        self, player_id: int, total_shots: int, successful_shots: int
    ) -> bool:
        if total_shots <= 0:
            logger.warning("PlayerStatsRepository: total_shots must be > 0.")
            return False

        stats = self.get_player_stats(player_id)
        if stats is None:
            logger.error("PlayerStatsRepository: Stats not found for player %d.", player_id)
            return False

        base = stats.average_accuracy * max(stats.total_battles - 1, 0)
        new_shot_accuracy = successful_shots / total_shots
        stats.average_accuracy = (base + new_shot_accuracy) / max(stats.total_battles, 1)
        return self._update_player_stats(stats)


    # ── LEADERBOARDS ───────────────────────────────────────────────────────────
    def get_top_players_by_level(self, top_count: int = 10) -> list[PlayerStats]:
        rows = self._db.execute_reader(
            "SELECT * FROM PlayerStats ORDER BY TotalWins DESC, TotalBattles DESC LIMIT ?",
            (top_count,),
        )
        return [_map_stats(row) for row in rows]

    def get_top_players_by_win_rate(self, top_count: int = 10) -> list[PlayerStats]:
        rows = self._db.execute_reader(
            """SELECT * FROM PlayerStats
               WHERE TotalBattles > 0
               ORDER BY (CAST(TotalWins AS FLOAT) / TotalBattles) DESC, TotalBattles DESC
               LIMIT ?""",
            (top_count,),
        )
        return [_map_stats(row) for row in rows]

    def get_player_rank(self, player_id: int, by_win_rate: bool = False) -> int:
        """Returns 1-based rank of the player (1 = best)."""
        if by_win_rate:
            count = self._db.execute_scalar(
                """SELECT COUNT(*) FROM PlayerStats
                   WHERE (CAST(TotalWins AS FLOAT) / NULLIF(TotalBattles, 0)) >
                         (SELECT CAST(TotalWins AS FLOAT) / NULLIF(TotalBattles, 0)
                          FROM PlayerStats WHERE PlayerID = ?)""",
                (player_id,),
            )
        else:                                                           # LIPSEA
            count = self._db.execute_scalar(
                """SELECT COUNT(*) FROM PlayerStats
                WHERE TotalWins > (SELECT TotalWins FROM PlayerStats WHERE PlayerID = ?)""",
                (player_id,),
            )
        return int(count or 0) + 1  # 1-indexed

    
    # ── RESET (admin / debug) ──────────────────────────────────────────────────
    def reset_player_stats(self, player_id: int) -> bool:
        result = self._db.execute_non_query(
            """UPDATE PlayerStats
            SET TotalBattles = 0,
                TotalWins = 0, TotalLosses = 0, TotalDraws = 0,
                AverageAccuracy = 0.0,
                TotalShipsDestroyed = 0, TotalShipsLost = 0,
                BattleshipKills = 0, BattleshipDeaths = 0, BattleshipAccuracy = 0,
                CruiserKills = 0, CruiserDeaths = 0, CruiserAccuracy = 0,
                DestroyerKills = 0, DestroyerDeaths = 0, DestroyerAccuracy = 0,
                CorvetteKills = 0, CorvetteDeaths = 0, CorvetteAccuracy = 0,
                TorpedoBoatKills = 0, TorpedoBoatDeaths = 0, TorpedoBoatAccuracy = 0,
                LastUpdated = ?
            WHERE PlayerID = ?""",
            (_now(), player_id),
        )
        if result > 0:
            logger.info("PlayerStatsRepository: Stats reset for player %d.", player_id)
        return result > 0

    # ── Private ────────────────────────────────────────────────────────────────

    def _update_player_stats(self, stats: PlayerStats) -> bool:
        result = self._db.execute_non_query(
            """UPDATE PlayerStats
               SET TotalBattles = ?,
                   TotalWins = ?, TotalLosses = ?, TotalDraws = ?,
                   AverageAccuracy = ?, TotalShipsDestroyed = ?, TotalShipsLost = ?,
                   LastUpdated = ?, BattleshipKills = ?, BattleshipDeaths = ?, BattleshipAccuracy = ?,
                    CruiserKills = ?, CruiserDeaths = ?, CruiserAccuracy = ?,
                    DestroyerKills = ?, DestroyerDeaths = ?, DestroyerAccuracy = ?,
                    CorvetteKills = ?, CorvetteDeaths = ?, CorvetteAccuracy = ?,
                    TorpedoBoatKills = ?, TorpedoBoatDeaths = ?, TorpedoBoatAccuracy = ?
               WHERE PlayerID = ?""",
            (
                stats.total_battles,
                stats.total_wins,
                stats.total_losses,
                stats.total_draws,
                stats.average_accuracy,
                stats.total_ships_destroyed,
                stats.total_ships_lost,
                _now(),
                stats.battleship_kills, stats.battleship_deaths, stats.battleship_accuracy,
                stats.cruiser_kills, stats.cruiser_deaths, stats.cruiser_accuracy,
                stats.destroyer_kills, stats.destroyer_deaths, stats.destroyer_accuracy,
                stats.corvette_kills, stats.corvette_deaths, stats.corvette_accuracy,
                stats.torpedo_boat_kills, stats.torpedo_boat_deaths, stats.torpedo_boat_accuracy,
                stats.player_id,
            ),
        )
        return result > 0

    def _insert_battle_history(self, player_id: int, battle: BattleRecord) -> None:
        self._db.execute_non_query(
            """INSERT INTO BattleHistory
               (PlayerID, OpponentName, BattleDate, Result,
                ShipsDestroyed, ShipsLost, DifficultyLevel, BattleMode)
               VALUES (?, ?, ?, ?, ?, ?, ?, ?)""",
            (
                player_id,
                battle.opponent_name,
                _now(),
                battle.result,
                battle.ships_destroyed,
                battle.ships_lost,
                battle.difficulty_level,
                battle.battle_mode,
            ),
        )


# ── Module-level functions ─────────────────────────────────────────────────────

def _now() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S")


def _emit(callbacks: list[Callable], *args) -> None:
    for cb in callbacks:
        try:
            cb(*args)
        except Exception as exc:
            logger.error("Signal callback raised an exception: %s", exc)


def _map_stats(row: dict) -> PlayerStats:
    return PlayerStats(
        stat_id=int(row.get("StatID", 0)),
        player_id=int(row.get("PlayerID", 0)),
        total_battles=int(row.get("TotalBattles", 0)),
        total_wins=int(row.get("TotalWins", 0)),
        total_losses=int(row.get("TotalLosses", 0)),
        total_draws=int(row.get("TotalDraws", 0)),
        average_accuracy=float(row.get("AverageAccuracy", 0.0)),
        total_ships_destroyed=int(row.get("TotalShipsDestroyed", 0)),
        total_ships_lost=int(row.get("TotalShipsLost", 0)),
        last_updated=str(row.get("LastUpdated", "") or ""),
        battleship_kills=int(row.get("BattleshipKills", 0)),      # NOU
        battleship_deaths=int(row.get("BattleshipDeaths", 0)),    # NOU
        battleship_accuracy=int(row.get("BattleshipAccuracy", 0)),# NOU
        cruiser_kills=int(row.get("CruiserKills", 0)),            # NOU
        cruiser_deaths=int(row.get("CruiserDeaths", 0)),          # NOU
        cruiser_accuracy=int(row.get("CruiserAccuracy", 0)),      # NOU
        destroyer_kills=int(row.get("DestroyerKills", 0)),        # NOU
        destroyer_deaths=int(row.get("DestroyerDeaths", 0)),      # NOU
        destroyer_accuracy=int(row.get("DestroyerAccuracy", 0)),  # NOU
        corvette_kills=int(row.get("CorvetteKills", 0)),          # NOU
        corvette_deaths=int(row.get("CorvetteDeaths", 0)),        # NOU
        corvette_accuracy=int(row.get("CorvetteAccuracy", 0)),    # NOU
        torpedo_boat_kills=int(row.get("TorpedoBoatKills", 0)),   # NOU
        torpedo_boat_deaths=int(row.get("TorpedoBoatDeaths", 0)), # NOU
        torpedo_boat_accuracy=int(row.get("TorpedoBoatAccuracy", 0)),
    )
